File: Algorithm/과제 13/food.py
def read_data(): #데이터 입력
    f = open("food.inp", "r")
    n = int(f.readline())
    target = list(map(int, f.readline().split()))
    infor = []
    for i in range(n):
        infor.append(list(map(int, f.readline().split())))
    return n, target, infor

def print_data(ans): #데이터 출력
    f = open("food.out", "w")
    f.write(ans)
    f.close()

def process(n, target, infor):
    best = 0
    for i in range(n):
        best += infor[i][4]
        
    result = branch(target, infor, "", n, best)
    ans = ""

    if result[2] == "-1":
        return "0"
    else:
        for i in range(1, n+1):
            if result[2][i-1] == "1":
                ans += str(i) + " "

    return ans


def branch(target, infor, tree, n, best):
    #더이상 추가할 재료가 없을때
    if len(tree) == n:
        current = cal(infor, tree)
        if match(target, current) == -1:
            return [-1, -1, "-1"]
        else:
            return [current[4], sum(current[:4]), tree]

    #첫번째 bound 조건
    #앞으로 무슨일이 있어도, 최소조건을 만족 못할경우 bound 시킨
    tmpTree = tree
    for i in range(n - len(tree)):
        tmpTree += "1"

    tmp = cal(infor, tmpTree)
    if match(target, tmp) == -1:
            return [-1, -1, "-1"]

    #두번째 bound 조건
    #앞으로 무슨일이 있어도, 기존의 최저가격을 넘어버릴 경우
    tmpTree = tree
    for i in range(n - len(tree)):
        tmpTree += "0"

    tmp = 0
    for i in range(n):
        if tmpTree[i] == 1:
            tmp += infor[i][4]
    if tmp >= best:
        return [-1, -1, "-1"]

    #세번째 bound 조건
    #이미 조건을 넘겼을때, 바로 반환해버림
    tmp = cal(infor, tmpTree)
    if match(target, tmp) == 1:
        return [tmp[4], sum(tmp[:4]), tmpTree]

        
    #bound 되지 않았을때, 아래의 과정을 수행한
    case1 = branch(target, infor, tree + "1", n, best)
    if case1[0] < best:
        best = case1[0]
    case2 = branch(target, infor, tree + "0", n, best)
    if case2[0] < best:
        best = case2[0]

    if case1[0] != -1 and case2[0] != -1:
        if case1[0] == case2[0]:
            if case1[1] < case2[1]:
                return case2
            return case1
        elif case1[0] <= case2[0]:
            return case1
        else:
            return case2
    elif case1[0] == -1 and case2[0] == -1:
        return case1
    elif case1[0] != -1:
        return case1
    else:
        return case2
    

def cal(infor, tree):
    ans = [0, 0, 0, 0, 0]
    for i in range(len(tree)):
        if tree[i] == "1":
            for j in range(5):
                ans[j] += infor[i][j]
    return ans

def match(target, current):
    for i in range(4):
        if target[i] > current[i]:
            return -1
    return 1
        


n, target, infor = read_data()
ans = process(n, target, infor)
print_data(ans)

File: Algorithm/과제 13/food.cpp
#include <iostream>
#include <fstream>
#include <string>
using namespace std;

//마라톤 예제 1번을 보고 참고했습니다!!

//전역변수
int target[4] = { 0, };
int infor[50][5];
int zeroPrice[50];

int n, best = 0, bestSum = 0;
string ans = "";
string path = "";

//문제풀이를 위한 함수
void branch();
int match(int* current);
int sum(int* arr);

int main(void) {
	ifstream in("food.inp");
	ofstream out("food.out");
	
	//입력을 받는 부분
	in >> n;

	for (int i = 0; i < 4; i++)
		in >> target[i];

	for (int i = 0; i < n; i++)
		for (int j = 0; j < 5; j++)
			in >> infor[i][j];

	for (int i = 0; i < n; i++) {
		best += infor[i][4];
		ans.push_back('0');
	}

	//가격이 0이고, 영양소는 존재하는 특수 케이스를 검출합니다
	for (int i = 0; i < n; i++) {
		if (infor[i][4] == 0) {
			bool trigger = false;
			for (int j = 0; j < 4; j++) {
				if (infor[i][j] != 0) {
					trigger = true;
					break;
				}
			}
			if (trigger == true)
				zeroPrice[i] = 1;
			else
				zeroPrice[i] = 0;
		}
		else
			zeroPrice[i] = 0;
	}

	//정답이 없는경우 먼저 걸러냅니다.
	int isAns[4] = { 0, 0, 0, 0 };
	for (int i = 0; i < n; i++) {
		for (int j = 0; j < 4; j++) {
			isAns[j] += infor[i][j];
		}
	}

	if (match(isAns) != 1) {
		cout << "-1" << endl;
		return 0;
	}

	//결과를 출력하는 부분
	branch();

	string output = "";
	for (int i = 1; i < n + 1; i++) {
		if (zeroPrice[i - 1] == 1)
			output += to_string(i) + " ";
		else if(ans[i-1] == '1')
			output += to_string(i) + " ";
	}

	out.write(output.c_str(), output.size());
	in.close();
	out.close();
	return 0;
}

void branch() {
	int pathLen = path.length();
	//더이상 추가할 재료가 없는경우
	//다시 말해서, 트리가 최대길이에 도달했을때 값을 반환합니다
	if (pathLen == n) {
		int current[5] = { 0, 0, 0, 0, 0 };
		for (int i = 0; i < n; i++) {
			if (path[i] == '1') {
				for (int j = 0; j < 5; j++)
					current[j] += infor[i][j];
			}
		}

		string temp = path;
		if (match(current) != -1) {
			if (current[4] < best) {
				ans = temp;
				best = current[4];
				bestSum = sum(current);
			}
			else if (current[4] == best && sum(current) > bestSum) {
				ans = temp;
				bestSum = sum(current);
			}
		}
		path.pop_back();
		return;
	}

	//첫번째 bound 조건
	//지금까지의 값들로도 목표값을 달성했다면
	//더 이상 탐색을 하지 않고, 값을 반환합니다
	int zero[5] = { 0, 0, 0, 0, 0 };
	for (int i = 0; i < pathLen; i++) {
		if (path[i] == '1') {
			for (int j = 0; j < 5; j++)
				zero[j] += infor[i][j];
		}
	}

	string temp = path;
	if (match(zero) == 1) {
		if (zero[4] < best) {
			for (int i = 0; i < n - pathLen; i++) {
				temp += "0";
			}
			ans = temp;
			best = zero[4];
			bestSum = sum(zero);
		}
		if (path.length() != 0) {
			path.pop_back();
		}
		return;
	}

	//두번째 bound 조건
	//현재까지의 best와 비교해서, 무슨일이 있어도 현재값이 더 좋지않다면
	//더 이상 탐색을 하지 않고, 값을 반환합니다
	if (zero[4] > best) {
		if (path.length() != 0) {
			path.pop_back();
		}
		return;
	}

	//bound되지 않았다면
	//좀 더 탐색을 진행합니다
	path.push_back('1');
	branch();
	path.push_back('0');
	branch();

	if(path.length() != 0)
		path.pop_back();
	return;
}

int match(int* current) {
	for (int i = 0; i < 4; i++) {
		if (target[i] > current[i])
			return -1;
	}
	return 1;
}

int sum(int* arr) {
	int result = 0;
	for (int i = 0; i < 4; i++) {
		result += arr[i];
	}
	return result;
}